security/tests/integration/istio_ca_node_agent_env.go

```diff
@@ -0,0 +1,142 @@
+	NodeAgentTestEnv struct {
+		framework.TestEnv
+		name      string
+		ClientSet *kubernetes.Clientset
+		NameSpace string
+		Hub       string
+		Tag       string
+	}
+)
+
+const (
+	istioCaWithGivenCertificate = "istio-ca-with-given-certificate"
+	nodeAgent                   = "node-agent"
+	nodeAgentService            = "node-agent-service"
+	podGroupPostfix             = "-pod-group"
+)
+
+// NewNodeAgentTestEnv creates the environment instance
+func NewNodeAgentTestEnv(name string, kubeconfig string, hub string, tag string) *NodeAgentTestEnv {
+	clientset, err := CreateClientset(kubeconfig)
+	if err != nil {
+		glog.Errorf("failed to initialize K8s client: %s\n", err)
+		return nil
+	}
+
+	namespace, err := createTestNamespace(clientset, testNamespacePrefix)
+	if err != nil {
+		return nil
+	}
+
+	return &NodeAgentTestEnv{
+		ClientSet: clientset,
+		name:      name,
+		NameSpace: namespace,
+		Hub:       hub,
+		Tag:       tag,
+	}
+}
+
+// GetName return environment ID
+func (env *NodeAgentTestEnv) GetName() string {
+	return env.name
+}
+
+// GetComponents is the key of a environment
+// It defines what components a environment contains.
+// Components will be stored in framework for start and stop
+func (env *NodeAgentTestEnv) GetComponents() []framework.Component {
+	return []framework.Component{
+		NewKubernetesPod(
+			env.ClientSet,
+			env.NameSpace,
+			istioCaWithGivenCertificate,
+			fmt.Sprintf("%v/istio-ca-test:%v", env.Hub, env.Tag),
+			[]string{},
+			[]string{},
+		),
+		NewKubernetesService(
+			env.ClientSet,
+			env.NameSpace,
+			"istio-ca",
+			v1.ServiceTypeClusterIP,
+			8060,
+			map[string]string{
+				"pod-group": istioCaWithGivenCertificate + podGroupPostfix,
+			},
+			map[string]string{
+				kube.KubeServiceAccountsOnVMAnnotation: "nodeagent.google.com",
+			},
+		),
+		NewKubernetesPod(
+			env.ClientSet,
+			env.NameSpace,
+			nodeAgent,
+			fmt.Sprintf("%v/node-agent-test:%v", env.Hub, env.Tag),
+			[]string{},
+			[]string{},
+		),
+		NewKubernetesService(
+			env.ClientSet,
+			env.NameSpace,
+			nodeAgentService,
+			v1.ServiceTypeLoadBalancer,
+			8080,
+			map[string]string{
+				"pod-group": nodeAgent + podGroupPostfix,
+			},
+			map[string]string{},
+		),
+	}
+}
+
+// Bringup doing general setup for environment level, not components.
+// Bringup() is called by framework.SetUp()
+func (env *NodeAgentTestEnv) Bringup() error {
+	return nil
+}
+
+// Cleanup clean everything created by this test environment, not component level
+// Cleanup() is being called in framework.TearDown()
+func (env *NodeAgentTestEnv) Cleanup() error {
+	glog.Infof("cleaning up environment...")
+	err := deleteTestNamespace(env.ClientSet, env.NameSpace)
+	if err != nil {
+		glog.Errorf("failed to delete namespace: %v error: %v", env.NameSpace, err)
+	}
+	return nil
+}
+
+// GetNodeAgentIPAddress returns the external LoadBalancer IP address of the service
+func (env *NodeAgentTestEnv) GetNodeAgentIPAddress() (string, error) {
```

Comment:
@yutongz 

We should add some discovery support in the integration test framework so people can find the component addresses from an environment in a uniform way.

Reply:
Agree. I think we can add the `status` concept into framework and component address can be a field.